Better Ruby/Bundler infrastructure

cstrahan opened the pull request:
This replaces the old Bundler infrastructure. The old setup required a ton of monkey-patches and convoluted acrobatics to get Bundler to purely install gems in a single prefix [1].

This takes a different approach: we use a smaller set of patches to allow us to install gems into their own prefix, and then create a symlink forest to present Bundler with one prefix, and then write out a bunch of Bundler-aware binstubs into the tree [2].

This should be much more robust and easier to maintain (there's still some trickiness, but that's inevitable with Bundler's design).

---
1. Bundler _requires_ that all gems exist in one prefix - it's an assumption that is pervasive in the Bundler codebase.
2. "binstub" is the what the Ruby community call their executable wrappers.

matthiasbeyer commented:
* Wouldn't this cause a complete rebuild of the ruby stuff in nixpkgs?
* Will this be merged before or after the next nixos release? Maybe cc @lethalman or someone?
* _Please_ provide more information in your commit message. At the moment, it is as helpful as "Changes" or "Fixes" - You could paste in the mail from the mailinglist, at least.

cstrahan commented:
> Wouldn't this cause a complete rebuild of the ruby stuff in nixpkgs?

Yes, granted we don't have too many ruby packages at the moment.

> Will this be merged before or after the next nixos release? Maybe cc @lethalman or someone?

Given that we haven't split off a new branch for the impending release, I'll wait. Do we have a time in mind to branch off? If it's a couple weeks out, it might be worth while getting this into the next release. I'll plan on waiting though, unless someone here gives the thumbs up.

> Please provide more information in your commit message. At the moment, it is as helpful as "Changes" or "Fixes" - You could paste in the mail from the mailinglist, at least.

Absolutely - I intended to amend the message before merging.

cstrahan commented:
I think the commit message should now be solid.

matthiasbeyer commented:
ACK on the commit msg.

dharaj commented:
May I ask what the status of this task is? The current status of the ruby infrastructure has been nothing but confusion and pain for me. I could not manage to get the process outlined here to work at all: http://nix-dev.science.uu.narkive.com/mqHApvU4/announcing-new-ruby-support

cstrahan commented:
@dharaj What sort of problems are you coming up against? You can let me know here, or feel free to ping me on #nixos (I'm **cstrahan** there).

zimbatm commented on pkgs/development/interpreters/ruby/gemconfig/default.nix:
```diff
@@ -129,10 +143,19 @@ in
     '';
+    # for whatever reason, the xapian-ruby gem leaves out the gemspec
+    # from the archive, so if we've extracted the source from a gem package,
+    # re-use the gemspec from the metadata.
```
I think it's the other way around. Gems are not supposed to package their own gemspec, it's only because they are careless that it happens. I would extract and force the gemspec to be the one that comes from the metadata.

zimbatm commented on pkgs/development/interpreters/ruby/gemconfig/default.nix:
```diff
@@ -42,10 +42,16 @@ in
     buildFlags = [
       "--with-system-v8=true"
     ];
+    # due to `git ls-files` ...
```
I think this wouldn't happen if the gemspec comes from the metadata. Not 100% sure but I think the metadata contains the full file list.

cstrahan replied:
I believe you're correct. I'll make that the default, and also fix up the other things you've pointed out. Thanks for reviewing this :).

zimbatm commented on pkgs/tools/audio/mpdcron/default.nix:
```diff
@@ -1,4 +1,4 @@
-{ stdenv, fetchgit, autoconf, automake, libtool, pkgconfig, glib, libdaemon, buildRubyGem
+{ stdenv, fetchgit, autoconf, automake, libtool, pkgconfig, glib, libdaemon,
 , mpd_clientlib, curl, sqlite, ruby, bundlerEnv, libnotify, pandoc }:
```
typo: double comma ?

zimbatm commented:
+1 for the new approach. It should solve having to re-build all the ruby dependencies whenever the gemset.nix is changing.

Have you had any thoughts on injecting runtime dependencies like openssl ? Right now it has to be done on the app level but I feel like it could be made part of the gem dependencies.

Dealing with rubygems and bundler is quite insane. I wish they where only used to host, resolve and build dependencies. If fact it's possible to go one step further and completely abandon rubygems and bundler at runtime. All the runtime needs is all the ruby lib paths in the RUBYLIB environment variable. That also makes startup faster because no gem "activation" is necessary.

cstrahan commented:
I'm going to make a couple fixes and then push this to staging this weekend.

cstrahan commented:
I've updated everyone's packages to use the new infrastructure, and everyone appears to be working.

I have one last thing I need to do: fix some bugs in Bundix. After that, this is going to staging.

I noticed that GitLab has been broken (segfault during build) for some time, and as much as I'd like to fix it, I'm not sure what needs to happen. I've opened a issue here: https://github.com/cowboyd/therubyracer/issues/391

zimbatm commented:
As a workaround you can remove therubyracer as a dependency and add nodejs
to the buildInput. Both are going to work with execjs and node is generally
more stable that having two VMs in the same executable. The dependency is
also probably only required during build to compile the assets.

On Tue, 20 Oct 2015 01:39 Charles Strahan <notifications@github.com> wrote:

> I've updated everyone's packages to use the new infrastructure, and
> everyone appears to be working.
>
> I have one last thing I need to do: fix some bugs in Bundix. After that,
> this is going to staging.
>
> I noticed that GitLab has been broken (segfault during build) for some
> time, and as much as I'd like to fix it, I'm not sure what needs to happen.
> I've opened a issue here: cowboyd/therubyracer#391
> <https://github.com/cowboyd/therubyracer/issues/391>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/8612#issuecomment-149386026>.
>

cstrahan committed:
ruby: new bundler infrastructure

This improves our Bundler integration (i.e. `bundlerEnv`).

Before describing the implementation differences, I'd like to point a
breaking change: buildRubyGem now expects `gemName` and `version` as
arguments, rather than a `name` attribute in the form of
"<gem-name>-<version>".

Now for the differences in implementation.

The previous implementation installed all gems at once in a single
derivation. This was made possible by using a set of monkey-patches to
prevent Bundler from downloading gems impurely, and to help Bundler
find and activate all required gems prior to installation. This had
several downsides:

* The patches were really hard to understand, and required subtle
  interaction with the rest of the build environment.
* A single install failure would cause the entire derivation to fail.

The new implementation takes a different approach: we install gems into
separate derivations, and then present Bundler with a symlink forest
thereof. This has a couple benefits over the existing approach:

* Fewer patches are required, with less interplay with the rest of the
  build environment.
* Changes to one gem no longer cause a rebuild of the entire dependency
  graph.
* Builds take 20% less time (using gitlab as a reference).

It's unfortunate that we still have to muck with Bundler's internals,
though it's unavoidable with the way that Bundler is currently designed.
There are a number improvements that could be made in Bundler that would
simplify our packaging story:

* Bundler requires all installed gems reside within the same prefix
  (GEM_HOME), unlike RubyGems which allows for multiple prefixes to
  be specified through GEM_PATH. It would be ideal if Bundler allowed
  for packages to be installed and sourced from multiple prefixes.
* Bundler installs git sources very differently from how RubyGems
  installs gem packages, and, unlike RubyGems, it doesn't provide a
  public interface (CLI or programmatic) to guide the installation of a
  single gem. We are presented with the options of either
  reimplementing a considerable portion Bundler, or patch and use parts
  of its internals; I choose the latter. Ideally, there would be a way
  to install gems from git sources in a manner similar to how we drive
  `gem` to install gem packages.
* When a bundled program is executed (via `bundle exec` or a
  binstub that does `require 'bundler/setup'`), the setup process reads
  the Gemfile.lock, activates the dependencies, re-serializes the lock
  file it read earlier, and then attempts to overwrite the Gemfile.lock
  if the contents aren't bit-identical. I think the reasoning is that
  by merely running an application with a newer version of Bundler, you'll
  automatically keep the Gemfile.lock up-to-date with any changes in the
  format. Unfortunately, that doesn't play well with any form of
  packaging, because bundler will immediately cause the application to
  abort when it attempts to write to the read-only Gemfile.lock in the
  store. We work around this by normalizing the Gemfile.lock with the
  version of Bundler that we'll use at runtime before we copy it into
  the store. This feels fragile, but it's the best we can do without
  changes upstream, or resorting to more delicate hacks.

With all of the challenges in using Bundler, one might wonder why we
can't just cut Bundler out of the picture and use RubyGems. After all,
Nix provides most of the isolation that Bundler is used for anyway.

The problem, however, is that almost every Rails application calls
`Bundler::require` at startup (by way of the default project templates).
Because bundler will then, by default, `require` each gem listed in the
Gemfile, Rails applications are almost always written such that none of
the source files explicitly require their dependencies. That leaves us
with two options: support and use Bundler, or maintain massive patches
for every Rails application that we package.

Closes #8612

domenkozar commented:
A few thoughts from Guix community: http://dthompson.us/ruby-on-guix.html

cstrahan commented:
Pushed to staging at https://github.com/NixOS/nixpkgs/commit/b6c06e216bb3bface40eb8ea6576b25f9b2971dd.

cstrahan commented:
> Have you had any thoughts on injecting runtime dependencies like openssl ? Right now it has to be done on the app level but I feel like it could be made part of the gem dependencies.

We should now have a pretty good story for handling dependencies ([example](https://github.com/NixOS/nixpkgs/blob/b6c06e216bb3bface40eb8ea6576b25f9b2971dd/pkgs/development/interpreters/ruby/gemconfig/default.nix#L103-L105)).

> Dealing with rubygems and bundler is quite insane. I wish they where only used to host, resolve and build dependencies. If fact it's possible to go one step further and completely abandon rubygems and bundler at runtime. All the runtime needs is all the ruby lib paths in the RUBYLIB environment variable. That also makes startup faster because no gem "activation" is necessary. (EDIT: rails is a different story because it depends on bundler specifically)

@copumpkin's idea was to create a mini-bundler - that is, implement just enough of the API to not break things, while most of the unnecessary logic could be chopped out. Something like that might be worth pursuing.

zimbatm commented:
Nice, nixos is getting better every day.

For the mini bundler, it makes me thing of `bundle install --standalone` which generates a `bundler/setup.rb` ruby file that just adds all the gems to the $LOAD_PATH when loaded. Most apps just require that file on top so all you have to do is invoke the app with `ruby -I$APP_ROOT/bundle` and everything works as usual. Obviously rails thinks it knows better and requires the whole bundler shebang so that's another story.

kevincox commented:
Are there docs on how to use this anywhere?
